### ISSUE TITLE
get_schema_fields does not pust request to filter class

### ISSUE THREAD

**Issue by DeKaN** [carltongibson/django-filter#687]
```
def some_queryset(request):
    user = request.user
    return SomeModel.objects.filter(user=user)

class SomeFilterSet(FilterSet):
    some_model = ModelChoiceFilter(name='some_model', queryset=some_queryset)
```

Request always is None for schema generating

**Comment by rpkilby** [comment]
It seems to me that the real issue is schema generation (probably) failing when using the *callable* behavior of `ModelChoiceFilter.queryset`, since no request object is provided. This would go against user expectations of having a valid request object passed to the callable. 

----

Taking a quick look, I'd argue that `None` is the correct value to assume here (that said, I don't use the schema generation feature and am not aware of all the nuances). 

A few points:
- Looking at the calling hierarchy, it seems that `request=None` is a common and valid case. The request argument to `SchemaGenerator.get_schema()` is optional, per the [docs](http://www.django-rest-framework.org/api-guide/schemas/#get_schemaself-request). 
- Even if a request object is provided, the `public` argument may [override](https://github.com/encode/django-rest-framework/blob/3.6.1/rest_framework/schemas.py#L303) the request to be `None`.

I think a potential solution would be:
- do not invoke the callable queryset during schema generation
- document that queryset callables should expect either a `None` value or a request object.

**Comment by rpkilby** [comment]
Hi @DeKaN - I've added an alternate approach at #691. Any thoughts?

**Comment by DeKaN** [comment]
@rpkilby I think, that request should be added at filter instantiation, because it may have custom add or hide filters for public api / current user.

I agree with adding `request may be None` to documentation for queryset callables

**Comment by rpkilby** [comment]
> because it may have custom add or hide filters for public api / current user. 

Maybe, but by this logic we should also provide the `queryset` and `request.query_params`. My thinking is that  `get_schema_fields()` is largely called in a different context than `filter_queryset()` and `to_html()`, so not instantiating the `filter_class` and using `base_filters` seems acceptable. Either that or we instantiate the `filter_class` and match the behavior of `filter_queryset()`/`to_html()`.

```python
filter_class.base_filters.items()

# or

filter_class(
    getattr(view.request, 'query_params', None), 
    queryset=view.get_queryset(),
    request=view.request,
).filters.items()
```

Want to weigh in @carltongibson? :)

**Comment by carltongibson** [comment]
On the phone, so just quickly: I'm inclined to say we don't want to do anything too sophisticated here — not as the default anyway. 

Schema generation is ALWAYS going to need manual input to capture complex cases.

If people want to override get_schema_fields (which is equivalent to DRFs advice on manually specifying the schema) I'm happy to make available whatever they need.

**Comment by carltongibson** [comment]
But I don't want to make the base implementation much more sophisticated than it currently is.

**Comment by carltongibson** [comment]
I'm going to merge #691 for v1.0.3 but I'm going to leave this (and #688) open for now. I'll milestone to resolve for v1.1. 

* I'm generally sympathetic to the idea that we may **want** a per-user variation. 
* Passing the `request` itself isn't an issue (it's a small change) but I'm not yet sure of the degree of dynamism we want/need to support.

**Comment by rpkilby** [comment]
Hi @carltongibson - both this and #688 can be closed, now that #691 is merged.
